opm/simulators/wells/StandardWellConnections.cpp

```diff
@@ -95,212 +207,257 @@ computeDensities(const std::vector<Scalar>& perfComponentRates,
+        const auto* const rho_s = &props.surf_dens_perf[perf*num_comp + 0];
+        const auto* const b     = &props.b_perf        [perf*num_comp + 0];
+
+        auto& rho = this->perf_densities_[perf];
 
-    // 1. Compute the flow (in surface volume units for each
-    //    component) exiting up the wellbore from each perforation,
-    //    taking into account flow from lower in the well, and
-    //    in/out-flow at each perforation.
-    std::vector<Scalar> q_out_perf((nperf)*num_comp, 0.0);
+        auto volrat = rho = Scalar{0};
+        for (auto comp = 0*num_comp; comp < num_comp; ++comp) {
+            rho    += previousMixture[comp] * rho_s[comp];
+            volrat += currentMixture [comp] / b    [comp];
+        }
 
-    // Step 1 depends on the order of the perforations. Hence we need to
-    // do the modifications globally.
-    // Create and get the global perforation information and do this sequentially
-    // on each process
+        rho /= volrat;
+    }
+}
+
+template <class FluidSystem, class Indices>
+std::vector<typename StandardWellConnections<FluidSystem, Indices>::Scalar>
+StandardWellConnections<FluidSystem, Indices>::
+calculatePerforationOutflow(const std::vector<Scalar>& perfComponentRates) const
+{
+    const int nperf    = this->well_.numPerfs();
+    const int num_comp = this->well_.numComponents();
+
+    auto q_out_perf = std::vector<Scalar>(nperf * num_comp, Scalar{0});
+
+    const auto& factory = this->well_.parallelWellInfo()
+        .getGlobalPerfContainerFactory();
 
-    const auto& factory = well_.parallelWellInfo().getGlobalPerfContainerFactory();
     auto global_q_out_perf = factory.createGlobal(q_out_perf, num_comp);
-    auto global_perf_comp_rates = factory.createGlobal(perfComponentRates, num_comp);
 
-    // TODO: investigate whether we should use the following techniques to calcuate the composition of flows in the wellbore
-    // Iterate over well perforations from bottom to top.
+    const auto global_perf_comp_rates = factory
+        .createGlobal(perfComponentRates, num_comp);
+
+    // TODO: investigate whether we should use the following techniques to
+    // calcuate the composition of flows in the wellbore.  Iterate over well
+    // perforations from bottom to top.
     for (int perf = factory.numGlobalPerfs() - 1; perf >= 0; --perf) {
         for (int component = 0; component < num_comp; ++component) {
-            auto index = perf * num_comp + component;
-            if (perf == factory.numGlobalPerfs() - 1) {
-                // This is the bottom perforation. No flow from below.
-                global_q_out_perf[index] = 0.0;
-            } else {
-                // Set equal to flow from below.
-                global_q_out_perf[index] = global_q_out_perf[index + num_comp];
-            }
+            const auto index = perf*num_comp + component;
+            auto& q_out = global_q_out_perf[index];
+
+            q_out = (perf == factory.numGlobalPerfs() - 1)
```

Comment:
The comments are lost here, I find them useful in understanding what is going on so they should be reinstated.

Reply:
Most of the original comments remain in the [`computeDensities()`](https://github.com/OPM/opm-simulators/blob/853fda9d4207e1dd36bf95b943490352141b105c/opm/simulators/wells/StandardWellConnections.cpp#L209) member function from which this function was extracted.  That said, I've pushed an update to annotate some of the statements here.